src/main/java/org/commcare/core/network/CommCareNetworkServiceGenerator.java

```diff
@@ -0,0 +1,79 @@
+package org.commcare.core.network;
+
+
+import org.commcare.util.LogTypes;
+import org.javarosa.core.services.Logger;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.X509TrustManager;
+
+import okhttp3.ConnectionSpec;
+import okhttp3.HttpUrl;
+import okhttp3.Interceptor;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.logging.HttpLoggingInterceptor;
+import retrofit2.Retrofit;
+
+/**
+ * Provides an instance of CommCareNetworkService.
+ * We have declared everything static in this class as we want to use the same objects (OkHttpClient, Retrofit, …) throughout the app
+ * to just open one socket connection that handles all the request and responses.
+ */
+
+public class CommCareNetworkServiceGenerator {
+
+    // Retrofit needs a base url to generate an instance but since our apis are fully dynamic it's not getting used.
+    private static final String BASE_URL = "https://www.commcarehq.org/";
+
+    private static Retrofit.Builder builder = new Retrofit.Builder().baseUrl(BASE_URL);
+
+    private static Interceptor redirectionInterceptor = new Interceptor() {
+        @Override
+        public Response intercept(Chain chain) throws IOException {
+            Request request = chain.request();
+            Response response = chain.proceed(request);
+            if (response.code() == 301) {
+                String newUrl = response.header("Location");
+                if (!isValidRedirect(request.url(), HttpUrl.parse(newUrl))) {
+                    Logger.log(LogTypes.TYPE_WARNING_NETWORK, "Invalid redirect from " + request.url().toString() + " to " + response.request().url().toString());
+                    throw new IOException("Invalid redirect from secure server to insecure server");
+                }
+            }
+            return response;
+        }
+    };
+
+    private static AuthenticationInterceptor authenticationInterceptor = new AuthenticationInterceptor();
+
+    private static OkHttpClient.Builder httpClient = new OkHttpClient.Builder()
+            .connectTimeout(ModernHttpRequester.CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+            .readTimeout(ModernHttpRequester.CONNECTION_SO_TIMEOUT, TimeUnit.MILLISECONDS)
+            .addNetworkInterceptor(redirectionInterceptor)
+            .addInterceptor(authenticationInterceptor)
+            .followRedirects(true);
+
+    private static Retrofit retrofit = builder.client(httpClient.build()).build();
+
+    public static CommCareNetworkService createCommCareNetworkService(final String credential) {
+        authenticationInterceptor.setCredential(credential);
+        return retrofit.create(CommCareNetworkService.class);
+    }
+
+    private static boolean isValidRedirect(HttpUrl url, HttpUrl newUrl) {
```

Comment:
Quick Note: We only implemented this by hand before because the Raw Java HTTP libraries didn't do it for us. If Retrofit will replicate the same behavior we don't need to keep this implemented internally.

Reply:
OkHTTP builder only allows us to set `followRedirects(boolean)` which allows for all redirects  and `followSslRedirects(boolean)` which can restrict HTTP -> HTTPS and HTTPS -> HTTP. So there is still no way to set only HTTP -> HTTPS and restrict HTTPS -> HTTP

Reply:
Good to know.

Out of curiosity, does it specifically strike you as an odd decision for us to restrict https->http only? A lot of this code was written in a time when we had to re-implement lots of "core" functionality for most systems (like digest auth) by hand, so there are some spots where we wrote this that are quite specific which may be less important than standardizing on normal industry behavior now that we are in a world with standard libraries again.

Reply:
Well it doesn't sound odd to me rather a good security practice. Though I am guessing most people block both HTTPS -> HTTP and HTTP -> HTTPS. Though no one seems to raise questions on approach we are following.